scipy/stats/_stats_py.py

```diff
@@ -4716,53 +4740,73 @@ def pearsonr(x, y, *, alternative='two-sided', method=None):
     than zero which happens in about half of the cases if we sample x and y.
 
     """
-    n = len(x)
-    if n != len(y):
-        raise ValueError('x and y must have the same length.')
+    x = np.asarray(x)
+    y = np.asarray(y)
+
+    if axis is None:
+        x = x.ravel()
+        y = y.ravel()
+        axis = -1
+
+    axis_int = int(axis)
+    if axis_int != axis:
+        raise ValueError('`axis` must be an integer.')
+    axis = axis_int
+
+    n = x.shape[axis]
+    if n != y.shape[axis]:
+        raise ValueError('`x` and `y` must have the same length along `axis`.')
 
     if n < 2:
-        raise ValueError('x and y must have length at least 2.')
+        raise ValueError('`x` and `y` must have length at least 2.')
 
-    x = np.asarray(x)
-    y = np.asarray(y)
+    try:
+        x, y = np.broadcast_arrays(x, y)
+    except ValueError as e:
+        message = '`x` and `y` must be broadcastable.'
+        raise ValueError(message) from e
+
+    x = np.moveaxis(x, axis, -1)
+    y = np.moveaxis(y, axis, -1)
+    axis = -1
 
     if (np.issubdtype(x.dtype, np.complexfloating)
             or np.issubdtype(y.dtype, np.complexfloating)):
         raise ValueError('This function does not support complex data')
 
     # If an input is constant, the correlation coefficient is not defined.
-    if (x == x[0]).all() or (y == y[0]).all():
+    const_x = (x == x[..., [0]]).all(axis=-1)
+    const_y = (y == y[..., [0]]).all(axis=-1)
+    const_xy = const_x | const_y
+    if np.any(const_xy):
         msg = ("An input array is constant; the correlation coefficient "
                "is not defined.")
         warnings.warn(stats.ConstantInputWarning(msg), stacklevel=2)
-        result = PearsonRResult(statistic=np.nan, pvalue=np.nan, n=n,
-                                alternative=alternative, x=x, y=y)
-        return result
 
     if isinstance(method, PermutationMethod):
-        def statistic(y):
-            statistic, _ = pearsonr(x, y, alternative=alternative)
+        def statistic(y, axis):
+            statistic, _ = pearsonr(x, y, axis=axis, alternative=alternative)
```

Comment:
Looks like this and a few similar code blocks don't get hit by the regular test suite. Not sure how important that is, but ran a local check just now.

Reply:
I think they're hit by `test_resampling_pvalue`, which is `xslow`ed. Probably doesn't could run every time now that this is natively vectorized. Would you prefer that the `xslow` be removed?